src/Lean/Meta/ForEachExpr.lean

```diff
@@ -8,51 +8,47 @@ import Lean.Util.MonadCache
 import Lean.Meta.Basic
 
 namespace Lean.Meta
-namespace ForEachExpr
 
-abbrev M := MonadCacheT Expr Unit MetaM
-
-mutual
-
-private partial def visitBinder (fn : Expr → MetaM Bool) : Array Expr → Nat → Expr → M Unit
-  | fvars, j, Expr.lam n d b c => do
-    let d := d.instantiateRevRange j fvars.size fvars
-    visit fn d
-    withLocalDecl n c d fun x =>
-      visitBinder fn (fvars.push x) j b
-  | fvars, j, Expr.forallE n d b c => do
-    let d := d.instantiateRevRange j fvars.size fvars
-    visit fn d
-    withLocalDecl n c d fun x =>
-      visitBinder fn (fvars.push x) j b
-  | fvars, j, Expr.letE n t v b _ => do
-    let t := t.instantiateRevRange j fvars.size fvars
-    visit fn t
-    let v := v.instantiateRevRange j fvars.size fvars
-    visit fn v
-    withLetDecl n t v fun x =>
-      visitBinder fn (fvars.push x) j b
-  | fvars, j, e => visit fn $ e.instantiateRevRange j fvars.size fvars
-
-partial def visit (fn : Expr → MetaM Bool) (e : Expr) : M Unit :=
-  checkCache e fun _ => do
-    if (← liftM (fn e)) then
-      match e with
-      | .forallE ..   => visitBinder fn #[] 0 e
-      | .lam ..       => visitBinder fn #[] 0 e
-      | .letE ..      => visitBinder fn #[] 0 e
-      | .app f a      => visit fn f; visit fn a
-      | .mdata _ b    => visit fn b
-      | .proj _ _ b   => visit fn b
-      | _             => return ()
-
-end
-
-end ForEachExpr
-
-/-- Similar to `Expr.forEach'`, but creates free variables whenever going inside of a binder. -/
-def forEachExpr' (e : Expr) (f : Expr → MetaM Bool) : MetaM Unit :=
-  ForEachExpr.visit f e |>.run
+/-- Similar to `Expr.forEach'`, but creates free variables whenever going inside of a binder.
+If the inner function returns `false`, deeper subexpressions will not be visited.
+ -/
+partial def forEachExpr' {m} [Monad m] [MonadLiftT MetaM m] [MonadControlT MetaM m]
+  (input : Expr)
+  (fn : Expr → m Bool)
+  : m Unit := do
+  let _ : STWorld IO.RealWorld m := ⟨⟩
+  let _ : MonadLiftT (ST IO.RealWorld) m := { monadLift := fun x => liftM (m := MetaM) (liftM (m := ST IO.RealWorld) x) }
+  let rec visit (e : Expr) : MonadCacheT Expr Unit m Unit :=
```

Comment:
In Transform.lean this is `MonadCacheT ExprStructEq Expr m Expr`. Should I be using ExprStructEq?

Reply:
Good question. To be on the safe side it is better to use `ExprStructEq`. 
The difference is that `ExprStructEq` takes binder names and annotations into account. 
That is, `forEachExpr'` may skip `fun {a} => a` because `fun a => a` was already visited. 
I didn't check all uses of `forEachExp'`, but even if this distinction is irrelevant in every single use in this code base, external users may not expect it. What do you think?

Reply:
I am going to merge as-is. We can make change this in another PR.